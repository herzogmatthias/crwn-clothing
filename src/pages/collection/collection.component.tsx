import * as React from "react";
import { RouteComponentProps } from "react-router-dom";
import {
  TitleContainer,
  CollectionPageContainer,
  ItemsContainer,
  CollectionItemContainer,
} from "./collection.styles";
import CollectionsContext from "../../contexts/collections/collections.context";

type ICollectionPageProps = RouteComponentProps<{ categoryId: string }>;

function CollectionPage({ match }: ICollectionPageProps) {
  const collections = React.useContext(CollectionsContext);
  const collection = collections[match.params.categoryId];
  const { title, items } = collection;
  return (
    <CollectionPageContainer>
      <TitleContainer>{title}</TitleContainer>
      <ItemsContainer>
        {items.map((item) => (
          <CollectionItemContainer
            key={item.id}
            item={item}
          ></CollectionItemContainer>
        ))}
      </ItemsContainer>
    </CollectionPageContainer>
  );
}

export default CollectionPage;
