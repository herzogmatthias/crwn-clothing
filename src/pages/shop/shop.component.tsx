import * as React from "react";
import CollectionsOverview from "../../components/collections-overview/collections-overview.component";
import { Route, RouteComponentProps } from "react-router-dom";
import CollectionPage from "../collection/collection.component";
import {
  firestore,
  convertCollectionsSnapshotToMap,
} from "../../firebase/firebase.utils";
import { updateCollections } from "../../redux/shop/shop.actions";
import { connect, ConnectedProps } from "react-redux";

type IShopPageProps = ConnectedProps<typeof connector> & RouteComponentProps;

function ShopPage({ match, updateCollections }: IShopPageProps) {
  let unsubscripeFromSnapshot = null;
  React.useEffect(() => {
    const collectionRef = firestore.collection("collections");
    collectionRef.onSnapshot(async (snapshot) => {
      const collectionsMap = convertCollectionsSnapshotToMap(snapshot);
      updateCollections(collectionsMap);
    });
    return () => {};
  }, []);

  return (
    <div className="shop-page">
      <Route
        exact
        path={`${match.path}`}
        component={CollectionsOverview}
      ></Route>
      <Route
        path={`${match.path}/:categoryId`}
        component={CollectionPage}
      ></Route>
    </div>
  );
}

const mapDispatchToProps = {
  updateCollections: updateCollections,
};

const connector = connect(null, mapDispatchToProps);
export default connect(null, mapDispatchToProps)(ShopPage);
