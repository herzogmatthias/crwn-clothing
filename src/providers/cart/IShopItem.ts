export interface IShopItem {
  id: number;
  name: string;
  imageUrl: string;
  price: number;
}
