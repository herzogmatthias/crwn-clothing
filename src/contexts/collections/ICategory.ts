import { IShopItem } from "../../providers/cart/IShopItem";

export interface ICategory {
  id: string | number;
  title: string;
  routeName: string;
  items: IShopItem[];
}
