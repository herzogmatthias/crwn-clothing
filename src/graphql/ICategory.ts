import { IShopItem } from "./IShopItem";

export interface ICategory {
  id: number;
  title: string;
  routeName: string;
  items: IShopItem[];
}
