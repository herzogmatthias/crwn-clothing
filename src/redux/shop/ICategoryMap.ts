import { ICategory } from "./ICategory";

export interface ICategoryMap {
  [key: string]: ICategory;
}
