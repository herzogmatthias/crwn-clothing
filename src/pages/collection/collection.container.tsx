import React from "react";
import { gql, ApolloCurrentQueryResult } from "apollo-boost";
import { RouteComponentProps } from "react-router-dom";
import { Query } from "react-apollo";
import { ICategory } from "../../redux/shop/ICategory";
import Spinner from "../../components/spinner/spinner.component";
import CollectionPage from "./collection.component";
import { GET_COLLECTION_BY_TITLE } from "../../graphql/queries";

interface ICollectionPageContainerProps
  extends RouteComponentProps<{ categoryId: string }> {}

const CollectionPageContainer = (props: ICollectionPageContainerProps) => {
  return (
    <Query
      query={GET_COLLECTION_BY_TITLE}
      variables={{ title: props.match.params.categoryId }}
    >
      {({
        loading,
        data,
      }: ApolloCurrentQueryResult<{ getCollectionsByTitle: ICategory }>) => {
        if (loading) return <Spinner></Spinner>;
        else {
          const { getCollectionsByTitle } = data!;
          return (
            <CollectionPage
              {...props}
              collection={getCollectionsByTitle}
            ></CollectionPage>
          );
        }
      }}
    </Query>
  );
};

export default CollectionPageContainer;
