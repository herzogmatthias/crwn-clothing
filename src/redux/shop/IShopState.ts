import { ICategoryMap } from "./ICategoryMap";

export interface IShopState {
  collections: ICategoryMap | null;
  isFetching: boolean;
  errorMessage: string;
}
