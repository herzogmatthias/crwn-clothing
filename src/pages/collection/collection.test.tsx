import React from "react";
import { shallow, ShallowWrapper } from "enzyme";

import { CollectionPage } from "./collection.component";
import { ICategory } from "../../redux/shop/ICategory";
import { IShopItem } from "../../redux/shop/IShopItem";
import { CollectionItemContainer } from "./collection.styles";
describe("NON CONNECTED COLLECTIONPAGE TESTS", () => {
  let wrapper: ShallowWrapper;
  let mockItems: IShopItem[] = [
    { id: 1, imageUrl: "asdfsd", name: "hat", price: 23 },
    { id: 2, imageUrl: "asdfsd", name: "trousers", price: 23 },
    { id: 3, imageUrl: "asdfsd", name: "tshirt", price: 23 },
  ];
  beforeEach(() => {
    const mockCollection: ICategory = {
      items: mockItems,
      title: "Test",
      id: "asdf213",
      routeName: "/hats",
    };
    wrapper = shallow(<CollectionPage collection={mockCollection} />);
  });

  it("should render the CollectionPage component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render the same number of CollectionItems as collection array", () => {
    expect(wrapper.find(CollectionItemContainer).length).toBe(mockItems.length);
  });
});
