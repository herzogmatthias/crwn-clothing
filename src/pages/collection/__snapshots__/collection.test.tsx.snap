// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionPage should render the CollectionPage component 1`] = `
<styled.div>
  <styled.h2>
    Test
  </styled.h2>
  <styled.div>
    <Styled(Connect(CollectionItem))
      item={
        Object {
          "id": 1,
          "imageUrl": "asdfsd",
          "name": "hat",
          "price": 23,
        }
      }
      key="1"
    />
    <Styled(Connect(CollectionItem))
      item={
        Object {
          "id": 2,
          "imageUrl": "asdfsd",
          "name": "trousers",
          "price": 23,
        }
      }
      key="2"
    />
    <Styled(Connect(CollectionItem))
      item={
        Object {
          "id": 3,
          "imageUrl": "asdfsd",
          "name": "tshirt",
          "price": 23,
        }
      }
      key="3"
    />
  </styled.div>
</styled.div>
`;
